File: tunnel-tunnelcli/src/main/java/com/tuuzed/tunnelcli/TunnelManifest.java
package com.tuuzed.tunnelcli;

import org.jetbrains.annotations.Nullable;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.LinkedHashMap;
import java.util.Map;

public final class TunnelManifest {

    private static final Map<String, String> manifest = new LinkedHashMap<>();

    static {
        URL url = TunnelManifest.class.getResource("/META-INF/MANIFEST.MF");
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(url.openStream()))) {
            String line;
            while ((line = reader.readLine()) != null) {
                int index = line.indexOf(":");
                if (index != -1) {
                    String k = line.substring(0, index).trim();
                    String v = line.substring(index + 1).trim();
                    manifest.put(k, v);
                }
            }
        } catch (IOException e) {
            // pass
        }
    }

    @Nullable
    public static String versionCode() {
        return manifest.get("Tunnel-VersionCode");
    }

    @Nullable
    public static String versionName() {
        return manifest.get("Tunnel-VersionName");
    }

    @Nullable
    public static String lastCommitSHA() {
        return manifest.get("Tunnel-LastCommitSHA");
    }

    @Nullable
    public static String lastCommitDate() {
        return manifest.get("Tunnel-LastCommitDate");
    }

    @Nullable
    public static String buildDate() {
        return manifest.get("Tunnel-BuildDate");
    }
}

File: tunnel2-t2cli/src/main/java/tunnel2/t2cli/TunnelManifest.kt
package tunnel2.t2cli

import java.io.IOException
import java.util.*

object TunnelManifest {
    private val manifest = LinkedHashMap<String, String>()

    init {
        val url = TunnelManifest::class.java.getResource("/META-INF/MANIFEST.MF")
        try {
            url.openStream().bufferedReader().use { reader ->
                while (true) {
                    val line = reader.readLine() ?: break
                    val index = line.indexOf(":")
                    if (index != -1) {
                        val k = line.substring(0, index).trim { it <= ' ' }
                        val v = line.substring(index + 1).trim { it <= ' ' }
                        manifest[k] = v
                    }
                }
            }
        } catch (e: IOException) {
            // pass
        }
    }

    val versionCode = manifest["Tunnel-VersionCode"]
    val versionName = manifest["Tunnel-VersionName"]
    val lastCommitSHA = manifest["Tunnel-LastCommitSHA"]
    val lastCommitDate = manifest["Tunnel-LastCommitDate"]
    val buildDate = manifest["Tunnel-BuildDate"]
}
